File: ng2-dart/lib/service/activity.service.dart
import 'dart:async';
import 'dart:convert';

import 'package:angular2/core.dart';
import 'package:http/http.dart';

import '../model/activity.dart';

@Injectable()
class ActivityService {

	static final _headers = { 'Content-Type' : 'application/json' };
	static const _restUrl = '/rest/activity/';

	final Client _http;

	ActivityService(this._http);

	Future<List<Activity>> getActivities() async {
		try {
			final response = await _http.get(_restUrl);
			final activitys = _extractData(response)
			.map((value) => new Activity.fromJson(value))
			.toList();

			return activitys;
		} catch (e) {
			throw _handleError(e);
		}
	}

	Future<Activity> getActivity(int id) async {
		var url = '$_restUrl${id}';
		try {
			final response = await _http.get(url);
			final activity = _extractData(response)
			.map((value) => new Activity.fromJson(value));

			return activity;
		} catch (e) {
			throw _handleError(e);
		}
	}

	Future<Activity> save(Activity activity) async {
		if (activity.id != null) {
			return this._put(activity);
		} else {
			return this._post(activity);
		}
	}	

	Future<Activity> delete(int id) async {
		var url = '$_restUrl${id}';
		try {
			final response = await _http.delete(url);
			final activity = _extractData(response)
			.map((value) => new Activity.fromJson(value));

			return activity;
		} catch (e) {
			throw _handleError(e);
		}
	}

	Future<Activity> _post(Activity activity) async {
		try {
     		final response = await _http.post(_restUrl,
         	headers: _headers, body: JSON.encode(activity));
      
      		return new Activity.fromJson(_extractData(response));
   		} catch (e) {
      		throw _handleError(e);
    	}
	}

	Future<Activity> _put(Activity activity) async {
		try {
     		final response = await _http.put(_restUrl,
         	headers: _headers, body: JSON.encode(activity));
      
      		return new Activity.fromJson(_extractData(response));
   		} catch (e) {
      		throw _handleError(e);
    	}
	}

	dynamic _extractData(Response resp) => JSON.decode(resp.body)['data'];

	Exception _handleError(dynamic e) {
    	print(e); // for demo purposes only
    	return new Exception('Server error; cause: $e');
  	}

}

File: ng2-ts/src/app/service/activity.service.ts
import { Injectable }    from '@angular/core';
import { Headers, Http } from '@angular/http';

import { Observable } from 'rxjs';

import { Activity } from '../model/activity';

@Injectable()
export class ActivityService {
    
    private restUrl = '/rest/activity/';
    
    constructor(private http: Http) {}
    
    getActivitiesForKudos(kudosId: number) {
        return this.http.get(this.restUrl + '/kudos/' + kudosId)
            .map( (response:any) => {
                if (response.status == 200) {
                    return response.json() as Activity[];
                } else {
                    return null;
                }
            });
    }

    getActivity(id: number) {
        let url = this.restUrl + id;
        return this.http.get(url)
            .map( (response:any) => {
                if (response.status == 200) {
                    return response.json() as Activity;
                } else {
                    return null;
                }
            });
    }
    
    save(activity: Activity) {
        if (activity.id) {
            return this.put(activity);
        }
        return this.post(activity);
    }
    
    delete(activity: Activity) {
        let headers = new Headers();
        headers.append('Content-Type', 'application/json');
        let url = this.restUrl + activity.id;
        return this.http
            .delete(url, {
                headers: headers
            })
            .map( (response:any) => {
                if (response.status == 200) {
                    return true;
                } else {
                    return false;
                }
            });
    }

    private post(activity: Activity) {
        let headers = new Headers({
            'Content-Type': 'application/json'
        });
        return this.http
            .post(this.restUrl, JSON.stringify(activity), {
                headers: headers
            })
            .map( (response:any) => {
                if (response.status == 200) {
                    return response.json() as Activity;
                } else {
                    return null;
                }
            });
    }

    private put(activity: Activity) {
        let headers = new Headers();
        headers.append('Content-Type', 'application/json');
        let url = this.restUrl;
        return this.http
            .put(url, JSON.stringify(activity), {
                headers: headers
            })
            .map( (response:any) => {
                if (response.status == 200) {
                    return response.json() as Activity;
                } else {
                    return null;
                }
            });
    }
}